File: BinarySearch/C/src/binarysearch.c
#include "binarysearch.h"
#include <stdio.h>

bool binary_search(int array[], int element, int start, int end) {
	if (start > end) return false;

	int middleIndex = start + (end - start) / 2;
	int middleValue = array[middleIndex];

	if (element == middleValue) return true;

	if (element > middleValue) {
		return binary_search(array, element, middleIndex + 1, end);
	} else if (element < middleValue) {
		return binary_search(array, element, start, middleIndex - 1);
	}

	return false;
}

File: BinarySearch/Python/src/binarysearch.py
#!/usr/bin/env python2.7
# -*- coding: utf-8 -*-
'''
	binarysearch.py

	Created by Ardalan Samimi on 2019-03-29
'''

def set_defaults(array, start, end):
	if start == None:
		start = 0
	if end == None:
		end = len(array)

	return start, end

def binary_search(array, element, start = None, end = None):
	start, end = set_defaults(array, start, end)

	middle = (start + end) / 2

	if (start > end or middle >= len(array)):
		return -1

	if (array[middle] == element):
		return middle
	elif (element < array[middle]):
		return binary_search(array, element, start, middle - 1)
	else:
		return binary_search(array, element, middle + 1, end)
